crates/replication2/src/coordinator/mod.rs

```diff
@@ -0,0 +1,286 @@
+                // return `None` indicating that quorum hasn't been reached.
+                (Some(a), Some(b)) => break Some(a).filter(|_| a == b),
+
+                _ => {}
+            };
+        };
+
+        let mut reconciled_response = None;
+
+        // NOTE: currently the set of operations that can be reconciled and the set
+        // of repairable operations do not intersect. Once it's no longer the case it
+        // needs to be handled properly here.
+        let quorum_response = match quorum_response_idx {
+            None if is_write => None,
+
+            // If we didn't get the quorum and this is a read operation, then try to reconcile
+            // the responses.
+            None => reconciliation::reconcile(operation, &responses[..RF])
+                .map(|out| reconciled_response = Some(Ok(out)))
+                .pipe(|_| reconciled_response.as_ref()),
+
+            Some(idx) => responses[idx as usize].as_ref(),
+        }
+        .unwrap_or_else(|| {
+            metrics::counter!("wcn_replication_coordinator_inconsistent_results").increment(1);
+            const { &Err(Error::new(ErrorKind::Internal)) }
+        });
+
+        let callback_fut = callback.send_result(quorum_response);
+
+        let mut read_repair = ReadRepair::new(operation, &responses, &primary_quorum);
+
+        let complete_replication_fut = async {
+            // Drive all futures to completion
+            while let Some((operator_idx, response)) = response_futures.next().await {
+                // If read repair is scheduled, check extra responses in case they also need to
+                // be repaired.
+                if let Some(repair) = &mut read_repair {
+                    repair.check_response(operator_idx, response);
+                }
+            }
+
+            if let Some(repair) = read_repair {
+                repair.run().await;
+            }
+        };
+
+        let (callback_result, _) = (callback_fut, complete_replication_fut).join().await;
+
+        callback_result
+    }
+
+    async fn execute(&self, operation: Operation<'_>) -> storage_api::Result<operation::Output> {
+        let (tx, rx) = oneshot::channel();
+
+        let this = self.clone();
+        let operation = operation.into_owned();
+        let handle = tokio::spawn(async move {
+            this.execute_callback(operation.into(), ResponseChannel(tx))
+                .await
+        });
+        drop(handle);
+
+        match rx.await {
+            Ok(resp) => resp,
+            Err(_) => {
+                tracing::warn!("Coordinator::execute_callback task cancelled");
+                Err(Error::new(ErrorKind::Internal))
+            }
+        }
+    }
+}
+
+async fn execute<N: StorageApi>(
+    operator: &NodeOperator<N>,
+    operation: &Operation<'_>,
+) -> storage_api::Result<operation::Output> {
+    let mut retries_left = operator.nodes.len();
```

Comment:
Is it possible to get an empty set of `nodes` here under any conditions? I suppose it's not, but since you're already returning a result here, maybe make sure that it's never empty? Otherwise it may panic with an overflow on subtraction.

Reply:
This is an invariant of `NodeOperator` and it's being checked in the respective constructor. `NodeOperator::next_node` can also panic, but this would be a bug.

It's worth a comment tho, I'll add one.

Reply:
Also the field needs to be private, which it isn't rn. Thanks for pointing out.